User Epoch PNL

devey opened the pull request:
# Task:

Add historical PNL for user under Dashboard. 

## Description

Use historical `PricePerShare` (per epoch), users' withdraws (in shares) and deposits (in `USDC` collateral) to calculate PNL over epoch and plot it.
 
## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Document update
- [ ] Quality of life improvement

## Checklist

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes generate no new warnings
- [ ] I have NATSPEC'd any new functions
- [ ] If appropriate, I have properly tested my new functionality

devey commented:
@MeanBoyCousin any idea why deploy on Vercel fails? I rebased on `rysk-alpha`

MeanBoyCousin commented:
> @MeanBoyCousin any idea why deploy on Vercel fails? I rebased on `rysk-alpha`

From the build logs, it looks like linting errors.

https://vercel.com/rysk-finance/rysk-alpha/4JqQniJhP9bc9rLMJA3E5xxJdoug#L32

devey commented:
> > @MeanBoyCousin any idea why deploy on Vercel fails? I rebased on `rysk-alpha`
> 
> From the build logs, it looks like linting errors.
> 
> https://vercel.com/rysk-finance/rysk-alpha/4JqQniJhP9bc9rLMJA3E5xxJdoug#L32

thanks, fixed

ugolino commented:
@devey this looks good to merge IMO

MeanBoyCousin commented on packages/front-end/src/components/dashboard/UserEpochPNL.tsx:
```diff
@@ -0,0 +1,366 @@
+        tempPNL.length > 0 && setHistoricalPNL(tempPNL);
+      },
+    }
+  );
```
There are a lot of checks to `data?.` inside here. Would be cleaner to wrap everything in an if statement and remove the need for extra checks.

MeanBoyCousin commented on packages/front-end/src/components/dashboard/UserEpochPNL.tsx:
```diff
@@ -0,0 +1,366 @@
+                        dataKey={({ pnl }) =>
+                          parseFloat(utils.formatUnits(pnl, DECIMALS.USDC))
+                        }
+                        // TODO access color throw Tailwind helpers
```
I looked into this. It doesn't seem do-able with the way the repo is structured as CRA won't reference files outside of the `src` folder. Something for the future with a better framework. 👍

MeanBoyCousin replied:
Yeah, that is what I looked into. Annoyingly, one of the caveats with create-react-app is that it won't let you reference any files that live outside the `src` folder. This means we can't use the `resolveConfig` for any of the code inside `src` without moving the Tailwind config. I also looked into this, but the big brains at Tailwind have decided that being able to specify where your config file lives aren't required. 🤷

MeanBoyCousin commented on packages/front-end/src/components/dashboard/UserEpochPNL.tsx:
```diff
@@ -0,0 +1,366 @@
+                        dataKey={({ change }) =>
+                          parseFloat(
+                            utils.formatUnits(change, DECIMALS.USDC)
+                          ).toFixed(2)
```
We should think about removing the use of `toFixed` in favour of numeric rounding as it can cause floating point errors. Not a big deal for now but worth considering going forward.

devey replied:
got it, should we just multiply and divide instead? but agree that we should write down a small doc on how to do operations on front-end

MeanBoyCousin commented on packages/front-end/src/components/dashboard/UserEpochPNL.tsx:
```diff
@@ -0,0 +1,366 @@
+                          return date.toLocaleString("default", {
+                            month: "short",
+                            day: "2-digit",
+                          });
```
Recommend using dayjs here. More extensible and readable.

MeanBoyCousin commented on packages/front-end/src/hooks/useUserPosition.ts:
```diff
@@ -71,10 +73,10 @@ export const useUserPosition = () => {
         // e6
         const sharesCurrentValue = amountInShares
-          .mul(latestEpochSharePrice)
+          .mul(latestEpochSharePrice) // withdrawal epoch share price for receipt epoch - 1
```
Nice docs updates, thanks!

devey commented:
@ugolino I made some fixes from @kjr217 feedback, good for me now but need to solve conflicts

MeanBoyCousin reviewed (changes requested):
I'm not super keen on the proliferation of massive React files in this code base. I appreciate that in a more traditional setup, a lot of the computational stuff would be done on a server and abstracted away from the front end, but I think that if this is the way it has to be, we need to try and compartmentalise components in a better fashion.

I'm still getting familiar with the graph but I'd like to know if there are ways to better abstract some of the logic into there. At the moment, I am seeing the graph as the model layer in a typical MVC setup and with this, I'd typically move as much of the computational methods (data-related working) onto model methods. Can custom methods be added to the graph or do we need a middle layer to act as a view? Perhaps this would be better with NextJS as we could use the API views to query the graph and do the complex logical updates and then simply query that from the front end and only have the data we require for the components. Interested to hear your thoughts.

As far as compartmentalising goes, it'd be great to see things like:
- types/interfaces extracted to their own file or into a higher level of the repo to be more reusable.
- GraphQL queries and logic moved into their own hook-style file, removing logic from the actual JSX.
- Util functions extracted out for re-usability.

Happy to discuss all of this! 😄

devey commented on packages/front-end/src/components/dashboard/UserEpochPNL.tsx:
```diff
@@ -0,0 +1,366 @@
+            displayType={"text"}
+            decimalScale={2}
+            suffix=" USDC"
+          />
```
nice, I looked into this but what's the difference compared to this one?

MeanBoyCousin replied:
The `NumberFormat` component will automatically return a `span` tag so we end up with a `span` inside a `p` tag. It's not a big deal as it is semantically valid, but best to avoid unnecessary nesting if we can. Using the prop mentioned means we can just have the `p` tag. 👍

devey replied:
got it

devey commented on packages/front-end/src/components/dashboard/UserEpochPNL.tsx:
```diff
@@ -0,0 +1,366 @@
+  }
+
+  return null;
+};
```
Yep, I think we only have two charts for now.
Should find some more stable library as I've had various annoying issues while using Recharts.
That's probably a good time to create more reusable parts but these first charts are MVPs.

MeanBoyCousin replied:
I have always used [chartjs](https://react-chartjs-2.js.org/) in the past with little problems so this might be worth a try. We need to look at the branding as well because the charts on there look nothing like the charts we have now.

devey commented on packages/front-end/src/components/dashboard/UserEpochPNL.tsx:
```diff
@@ -0,0 +1,366 @@
+          className={"text-white"}
+          y={y}
+          dy={15}
+          dx={value > 0 ? -50 : 50}
```
This is the value coming from `dataKey` which in this case is:
```js
	dataKey={({ change }) =>
            parseFloat(utils.formatUnits(change, DECIMALS.USDC))
 	}
```

MeanBoyCousin replied:
So in that case, should the early return above be `value === 0` or just `!value`?

devey commented on packages/front-end/src/components/dashboard/UserEpochPNL.tsx:
```diff
@@ -0,0 +1,366 @@
+          y={y}
+          dy={15}
+          dx={value > 0 ? -50 : 50}
+          fill={value > 0 ? "green" : "red"}
```
Should also be a `Number`.

devey commented on packages/front-end/src/components/dashboard/UserEpochPNL.tsx:
```diff
@@ -0,0 +1,366 @@
+          dy={15}
+          dx={value > 0 ? -50 : 50}
+          fill={value > 0 ? "green" : "red"}
+          fontSize={13}
```
14 works 🙂

devey commented on packages/front-end/src/components/dashboard/UserEpochPNL.tsx:
```diff
@@ -0,0 +1,366 @@
+  // the users' withdrawal receipt holds the epoch number (that matches when it was initiated + 1)
+  // so it will be processed at that price per share, not latest
+  // and there is no way for user to reverse the withdrawal action
+  useQuery(
```
We would probably need to generate types from The Graph's schema (https://www.apollographql.com/docs/react/development-testing/static-typing/) to fix this here and in all other `useQuery`

MeanBoyCousin replied:
Possibly a time saver, but not a blocker to doing it manually? I did it with [this query.](https://github.com/rysk-finance/dynamic-hedging/blob/rysk-alpha/packages/front-end/src/components/VaultPerformance/VaultPerformance.tsx#L20)

devey commented on packages/front-end/src/components/dashboard/UserEpochPNL.tsx:
```diff
@@ -0,0 +1,366 @@
+                timestamp: ppsEpoch.timestamp,
+                dateLocale: dateLocale,
+                epoch: ppsEpoch.id,
+              });
```
i needed to read previous shares so made more sense to just have an array where i push epochs as i go and use map only to iterate over all epochs

MeanBoyCousin replied:
Understandable, but I'd recommend avoiding mutation wherever possible. Mutation generally leads to harder-to-debug and harder-to-follow code.

devey replied:
I don't think I'm mutating anything here as the graph variable in untouched, there is a new Array with this interface of data that is computed from graph data:
```ts
interface PNL {
  epoch: string;
  pnl: string;
  change: string;
  shares: string;
  timestamp: string;
  dateLocale: string;
}
```

MeanBoyCousin replied:
You are mutating `tempPNL` with each push. The map is constructing the list in memory and then you are pushing results from inside the map to another memory allocation and mutating it each time. Not only will just retuning improve readability but it'll also improve memory allocation. Win-win! :)

MeanBoyCousin commented on packages/front-end/src/components/dashboard/UserVault.tsx:
```diff
@@ -29,8 +29,7 @@ export const UserVault = () => {
   );
-  const [unredeemableCollateral, setUnredeemableCollateral] =
-    useState<BigNumber>(BigNumber.from(0));
+  const [unredeemableCollateral] = useState<BigNumber>(BigNumber.from(0));
```
Do you think we will ever need the setter here?

devey replied:
no, these are not necessary here as they are already being shown in `PositionTooltip` (which you access using the info button next to _Your Position_). For information, unredeemable collateral is collateral that is waiting to be deposited in the next epoch.

devey replied:
also just checked and seems like these tooltips are broken in `rysk-alpha`, going to open an issue

MeanBoyCousin reviewed (approved):
LGTM. Just have one question.

devey commented:
> LGTM. Just have one question.

nice, feel free to merge it when you can

ugolino commented:
LGTM. merging it
